usr/lib/common/asn1.c

```diff
@@ -3473,3 +3574,806 @@ CK_RV ber_decode_DHPublicKey(CK_BYTE *data,
 
     return rc;
 }
+
+/**
+ * An IBM Dilithium public key is given by:
+ *
+ *  SEQUENCE (2 elem)
+ *    SEQUENCE (2 elem)
+ *      OBJECT IDENTIFIER 1.3.6.1.4.1.2.267.1.6.5
+ *      NULL
+ *    BIT STRING (1 elem)
+ *      SEQUENCE (2 elem)
+ *        BIT STRING (256 bit)   = 32 bytes
+ *        BIT STRING (13824 bit) = 1728 bytes
+ */
+CK_RV ber_encode_IBM_DilithiumPublicKey(CK_BBOOL length_only,
+                          CK_BYTE **data, CK_ULONG *data_len,
+                          CK_ATTRIBUTE *rho, CK_ATTRIBUTE *t1)
+{
+    CK_ULONG len, offset, total, total_len;
+    CK_RV rc;
+    CK_BYTE *buf = NULL;
+    CK_BYTE *buf2 = NULL;
+    CK_BYTE *buf3 = NULL;
+    BerValue *val;
+    BerElement *ber;
+
+    UNUSED(length_only);
+
+    offset = 0;
+    rc = 0;
+    total_len = ber_AlgIdDilithiumLen;
+    total = 0;
+
+    /* Calculate storage for inner sequence */
+    rc |= ber_encode_INTEGER(TRUE, NULL, &len, NULL, rho->ulValueLen);
+    offset += len;
+    rc |= ber_encode_INTEGER(TRUE, NULL, &len, NULL, t1->ulValueLen);
+    offset += len;
+
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("%s ber_encode_Int failed with rc=0x%lx\n", __func__, rc);
+        return rc;
+    }
+
+    /* Allocate storage for inner sequence */
+    buf = (CK_BYTE *) malloc(offset);
+    if (!buf) {
+        TRACE_ERROR("%s Memory allocation failed\n", __func__);
+        return CKR_HOST_MEMORY;
+    }
+
+    /**
+     *    SEQUENCE (2 elem)
+     *       BIT STRING -> rho
+     *       BIT STRING -> t
+     */
+    offset = 0;
+    rc = ber_encode_BIT_STRING(FALSE, &buf2, &len,
+                            rho->pValue, rho->ulValueLen, 0);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("%s ber_encode_Int failed with rc=0x%lx\n", __func__, rc);
+        return rc;
+    }
+    memcpy(buf + offset, buf2, len);
+    offset += len;
+    free(buf2);
+
+    rc = ber_encode_BIT_STRING(FALSE, &buf2, &len,
+                            t1->pValue, t1->ulValueLen, 0);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("%s ber_encode_Int failed with rc=0x%lx\n", __func__, rc);
+        return rc;
+    }
+    memcpy(buf + offset, buf2, len);
+    offset += len;
+    free(buf2);
+
+    rc = ber_encode_SEQUENCE(FALSE, &buf2, &len, buf, offset);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("%s ber_encode_Seq failed with rc=0x%lx\n", __func__, rc);
+        return rc;
+    }
+
+    /* Calculate length of outer sequence */
+    rc = ber_encode_OCTET_STRING(TRUE, NULL, &total, buf2, len);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("%s ber_encode_Oct_Str failed with rc=0x%lx\n", __func__,
```

Comment:
for buf, see above.
buf2 needs not to be freed here, because ber_encode_OCTET_STRING is called to retrieve the length of the sequence. In this case no buf is allocated.